csrc/fusion_segmenter.cpp

```diff
@@ -3123,34 +3261,14 @@ void SegmentCandidateFinder::findSegments() {
 
       resetLevels();
 
-      for (auto& group : groups()) {
-        if (group->merged_ || group->isFusionInputGroup()) {
-          continue;
-        }
-        auto candidates = group->getMergeCandidates();
-        if (candidates.empty()) {
-          continue;
-        }
-
-        auto candidate_it = candidates.begin();
-        while (candidate_it != candidates.end() &&
-               !codeGenSupportedMerge(group, candidate_it->group)) {
-          candidate_it++;
-        }
-        if (candidate_it == candidates.end()) {
-          continue;
-        }
-
-        to_merge_.emplace_back(group);
-        to_merge_.emplace_back(candidate_it->group);
-
-        group->merged_ = true;
-        group->merge_with_ = candidate_it->group;
-        group->merge_through_ = candidate_it->edge;
+      // Try preferred merge first
+      for (auto& [group, neighbor] :
+           PreferredMergeCandidatePicker::get(groups())) {
+        trySetUpMerge(group, {neighbor});
+      }
 
-        candidate_it->group->merged_ = true;
-        candidate_it->group->merge_with_ = group;
-        candidate_it->group->merge_through_ = candidate_it->edge;
+      for (auto& group : groups()) {
+        trySetUpMerge(group);
```

Comment:
Should we skip this if `to_merge_` is not empty here, so that we actually apply the preferred merge and reevaluate if there are new preferred merges? I know that this won't make a difference today, but if in the future we want to generalize `PreferredMergeCandidatePicker` for other cases, I think this behavior makes more sense.